components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java

```diff
@@ -1294,39 +1286,28 @@ private void updateArtifactCount(String businessRuleUUID, int artifactCount)
     private void resolveInstanceCount(BusinessRule businessRule)
             throws TemplateManagerServiceException, TemplateInstanceCountViolationException {
         int count;
-        int deploymentStatus = 0;
-        try {
-            deploymentStatus = queryExecutor.executeRetrieveDeploymentStatus(businessRule.getUuid());
-        } catch (BusinessRulesDatasourceException e) {
-            throw new TemplateManagerServiceException("Failed to retrieve the status of the business rule '" +
-                    businessRule.getUuid() + "'");
-        }
-        if ((deploymentStatus != TemplateManagerConstants.DEPLOYED) &&
-                (deploymentStatus != TemplateManagerConstants.PARTIALLY_DEPLOYED)) {
-            if (businessRule instanceof BusinessRuleFromTemplate) {
-                BusinessRuleFromTemplate businessRuleFromTemplate = (BusinessRuleFromTemplate) businessRule;
-                String instanceLimit = getRuleTemplate(businessRuleFromTemplate).getInstanceCount();
-                count = getInstanceCount(businessRuleFromTemplate.getRuleTemplateUUID());
+        if (businessRule instanceof BusinessRuleFromTemplate) {
+            BusinessRuleFromTemplate businessRuleFromTemplate = (BusinessRuleFromTemplate) businessRule;
+            String instanceLimit = getRuleTemplate(businessRuleFromTemplate).getInstanceCount();
+            count = getInstanceCount(businessRuleFromTemplate.getRuleTemplateUUID());
+            if ("one".equalsIgnoreCase(instanceLimit) && count >= 1) {
+                throw new TemplateInstanceCountViolationException("Rule Template '" +
+                        ((BusinessRuleFromTemplate) businessRule).getRuleTemplateUUID() + "' can be instantiated " +
+                        "only once and has already been instantiated.");
+            }
+        } else if (businessRule instanceof BusinessRuleFromScratch) {
+            BusinessRuleFromScratch businessRuleFromScratch = (BusinessRuleFromScratch) businessRule;
+            List<RuleTemplate> ruleTemplates =
+                    (List<RuleTemplate>) getInputOutputRuleTemplates(businessRuleFromScratch);
```

Comment:
Why casting here? Does this method return generics. Cant we make the return specific?

Reply:
Fixed with PR #1205